lib/elixir_analyzer/test_suite/dna_encoding.ex

```diff
@@ -29,4 +30,113 @@ defmodule ElixirAnalyzer.TestSuite.DNAEncoding do
     called_fn name: :for
     comment Constants.dna_encoding_use_recursion()
   end
+
+  check_source "uses tail call recursion" do
+    type :essential
+    suppress_if "does not call any Enum functions", :fail
+    suppress_if "does not call any Stream functions", :fail
+    suppress_if "does not call any List functions", :fail
+    suppress_if "does not use list comprehensions", :fail
+    comment Constants.dna_encoding_use_tail_call_recursion()
+
+    check(%Source{code_ast: code_ast}) do
+      {_, tail_call_recursive_functions} =
+        Macro.prewalk(
+          code_ast,
+          [],
+          fn node, acc -> find_tail_call_recursive_functions(node, acc) end
+        )
+
+      tail_call_recursive_functions = Enum.uniq(tail_call_recursive_functions)
+
+      {_, all_recursive_functions} =
+        Macro.prewalk(
+          code_ast,
+          [],
+          fn node, acc -> find_all_recursive_functions(node, acc) end
+        )
+
+      all_recursive_functions = Enum.uniq(all_recursive_functions)
+
+      non_tail_call_recursive_functions = all_recursive_functions -- tail_call_recursive_functions
+
+      if non_tail_call_recursive_functions != [] || tail_call_recursive_functions == [] do
+        false
+      else
+        true
+      end
```

Comment:
@jiegillet I'm feeling a bit silly asking this since I've been working on this project for so long... but how can I return a custom `%Comment{}` struct here? I would like to pass the two function lists to the comment template.

Reply:
Not silly at all, I also had no idea, even though I wrote `check_source` haha.

Looking at `ElixirAnalyzer.ExerciseTest.CheckSource.Compiler` it looks like the comment is created from the parameters `name`, `comment`, `type` and `suppress_if`. But looking at `ElixirAnalyzer.Comment` it seems there is one parameter missing: the map `params`, which is what you are looking for to build a custom comment.

`params` can't be added the same way since it's produced in the function, but maybe we can change the compiler to something like (untested)

```elixir
    test_description =
      Macro.escape(fn params -> %Comment{
        name: name,
        comment: comment,
        type: type,
        suppress_if: suppress_if,
        params: params
      })  

    quote do
      (fn %Source{} = source ->
         case unquote(check_function).(source) do
           false-> {:fail, unquote(test_description).(nil)}
           true-> {:pass, unquote(test_description).(nil)}
           params when is_map(params) -> {:pass, unquote(test_description).(params)}
           _ -> raise "check must be true, false or a map of parameters for the comment"
         end 
       end).(unquote(code_source))
    end 
```

Reply:
Thank you, that was close to it!